[Feature][Issue-2214] Support to select user when binding member.

RocMarshal opened the pull request:
<!--
Thank you for contributing to StreamPark! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

## Contribution Checklist

  - If this is your first time, please read our contributor guidelines: [Submit Code](https://streampark.apache.org/community/submit_guide/submit_code).

  - Make sure that the pull request corresponds to a [GITHUB issue](https://github.com/apache/incubator-streampark/issues).

  - Name the pull request in the form "[Feature] Title of the pull request", where *Feature* can be replaced by `Hotfix`, `Bug`, etc.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.

  - If the PR is unfinished, add `[WIP]` in your PR title, e.g., `[WIP][Feature] Title of the pull request`.

-->

## What changes were proposed in this pull request

Issue Number: close #2214 
<!--(For example: This pull request proposed to add checkstyle plugin).-->

## Brief change log

[Feature][Issue-2214] Team add number, can direct select.

## Verifying this change

<!--*(Please pick either of the following options)*-->

- *Manually verified the change by testing locally.* 

## Does this pull request potentially affect one of the following parts
 - Dependencies (does it add or upgrade a dependency): (yes / **no**)

RocMarshal committed:
[Feature][Issue-2214] Support to select user when binding member.

Add 'findUsersNotInTeam' interface for backend.

RocMarshal committed:
[Feature][Issue-2214] Support to select user when binding member.

Add selection-options part view and logic for FE.

RocMarshal commented:
Hi, @wangsizhu0504 .
The new interface `candidateUsers` is healthy.
The data queried from backend can't be rendered in FE, Could you help to check it ? Thank you in advance.
CC @1996fanrui @wangsizhu0504 @wolfboys

RocMarshal commented:
<img width="1413" alt="截屏2023-01-19 19 57 52" src="https://user-images.githubusercontent.com/64569824/213438130-7084c38f-5fb9-4435-b5fb-8c512b610033.png">
<img width="1517" alt="截屏2023-01-19 19 59 14" src="https://user-images.githubusercontent.com/64569824/213438205-592e58cc-1902-434e-b36c-ad6ba0acd0a3.png">

RocMarshal committed:
[Feature][Issue-2214] Support to select user when binding member.

Fix the render logic for userOptions

RocMarshal commented:
Thanks for the help from @1996fanrui @wangsizhu0504 offline.
I fixed the little case in https://github.com/apache/incubator-streampark/pull/2262/commits/c42061c83a38a4738fdcc001c6fef5db50252c7b
Could you @wolfboys help to review it ? thx

wolfboys commented:
> Thanks for the help from @1996fanrui @wangsizhu0504 offline. I fixed the little case in [c42061c](https://github.com/apache/incubator-streampark/commit/c42061c83a38a4738fdcc001c6fef5db50252c7b) Could you @wolfboys help to review it ? thx

Thanks for your contribution, happy chinese new year👏, i'll review it later

1996fanrui commented on streampark-console/streampark-console-service/src/main/resources/mapper/system/MemberMapper.xml:
```diff
@@ -68,4 +68,15 @@
         </if>
     </select>
 
+    <select id="findUsersNotInTeam" resultType="org.apache.streampark.console.system.entity.User">
+        select tu.* from t_user tu
+        where (tu.user_id, tu.username)
+        not in (
+                    select u.user_id, u.username
+                        from t_user u join t_member m
+                        on m.team_id = #{teamId}
+                        and m.user_id = u.user_id
+                )
+    </select>
```
Can the `username` be removed?

1996fanrui reviewed (changes requested):
Hi @RocMarshal , thanks for your contribution. I have tested this feature on My Mac, it works well.

Mostly looks good, I left a minor comment, please help take a look in your free time, thanks~

BTW, the PR title and the commit message need to be changed. How about `Support to select user when binding member.`?

RocMarshal commented:
> 

The title `Support to select user when binding member.` SGTM.

RocMarshal committed:
Update based on review comments.

Update commit message;
Simplify the sql statement.

RocMarshal committed:
Update candidate users options in time

RocMarshal committed:
Update based on review comments.

Update the return-type of the interface.

Fix the code style for changed files in FE.

wolfboys commented on streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/system/controller/MemberController.java:
```diff
@@ -53,6 +54,12 @@ public RestResponse memberList(RestRequest restRequest, Member member) {
 
+  @PostMapping("candidateUsers")
+  public RestResponse candidateUsers(RestRequest restRequest, Long teamId) {
+    IPage<User> userList = memberService.findCandidateUsers(teamId, restRequest);
```
I suggest return `List<User>` instead of `IPage<User>`, The fact is that this does not use the paging function

RocMarshal replied:
@wolfboys thanks for your review & suggestions. 
The truth what you described is right.  
Before reaching a consensus, I want to express the original intention of this design:
The `interface implemented with Pagination`  is a superset of `interface implemented without Pagination` , In other words, if there are other calls that need paging, the interface can still be reused. It will lead to some redundancy in the current front-end calling.

Looking forward to the final conclusion from you.

wolfboys replied:
> @wolfboys thanks for your review & suggestions. The truth what you described is right. Before reaching a consensus, I want to express the original intention of this design: The `interface implemented with Pagination` is a superset of `interface implemented without Pagination` , In other words, if there are other calls that need paging, the interface can still be reused. It will lead to some redundancy in the current front-end calling.
> 
> Looking forward to the final conclusion from you.

Happy Chinese new year, thanks for your feedback, My point : This method does not seem to be reusable. That is, it seems to be customized for specific requests, not a generic method (find user not in team),with explicit query conditions, what you think?

RocMarshal replied:
> > @wolfboys thanks for your review & suggestions. The truth what you described is right. Before reaching a consensus, I want to express the original intention of this design: The `interface implemented with Pagination` is a superset of `interface implemented without Pagination` , In other words, if there are other calls that need paging, the interface can still be reused. It will lead to some redundancy in the current front-end calling.
> > Looking forward to the final conclusion from you.
> 
> Happy Chinese new year, thanks for your feedback, My point : This method does not seem to be reusable. That is, it seems to be customized for specific requests, not a generic method (find user not in team),with explicit query conditions, what you think?

Got it & updated it.
Happy Chinese new year~ & Thanks for the clarification.

wolfboys commented on streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/system/service/impl/MemberServiceImpl.java:
```diff
@@ -82,6 +82,14 @@ public IPage<Member> findUsers(Member member, RestRequest request) {
+    Page<User> page = new Page<>();
+    page.setCurrent(request.getPageNum());
+    page.setSize(request.getPageSize());
+    return baseMapper.findUsersNotInTeam(page, teamId);
```
Pagination will add unnecessary trouble, right?

wolfboys commented on streampark-console/streampark-console-webapp/src/views/system/member/MemberDrawer.vue:
```diff
@@ -99,8 +101,17 @@
+        componentProps: {
+          disabled: unref(isUpdate),
+          api: fetchCandidateUsers,
+          params: { page: 1, pageSize: 9999, teamId: userStore.getTeamId},
```
Pagination will add unnecessary trouble, right?

wolfboys reviewed (commented):
Hi @RocMarshal , thanks for your hard work. I have tested, it works well.
There is a small suggestion, we can not use pagination.

RocMarshal commented on streampark-console/streampark-console-webapp/src/views/system/member/MemberDrawer.vue:
```diff
@@ -41,7 +43,7 @@
   import { RuleObject } from 'ant-design-vue/lib/form';
   import { StoreValue } from 'ant-design-vue/lib/form/interface';
-  import { fetchAddMember, fetchCheckUserName, fetchUpdateMember } from '/@/api/system/member';
+  import { fetchAddMember, fetchCandidateUsers, fetchCheckUserName, fetchUpdateMember } from "/@/api/system/member";
```
keep the same code-style. `'`

RocMarshal commented on streampark-console/streampark-console-webapp/src/views/system/member/MemberDrawer.vue:
```diff
@@ -53,7 +55,7 @@
       type: Array as PropType<Array<Partial<RoleListItem>>>,
       default: () => [],
-    },
+    }
```
```suggestion
    },
```

wolfboys reviewed (approved):
LGTM, thanks for your hard work.👏 Happy Chinese new year!

1996fanrui reviewed (approved):
Thanks for the update and review.

LGTM.
